### PR TITLE
[jit] Fix the class of the stack entry for ldelema, it should be the class not its element class.

### PR DESCRIPTION
**Pull request by vargaz**
Fixes https://github.com/mono/mono/issues/10101.

### CODE REVIEW

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -3976,7 +3976,7 @@ mini_emit_ldelema_1_ins (MonoCompile *cfg, MonoClass *klass, MonoInst *arr, Mono
 		static const int fast_log2 [] = { 1, 0, 1, -1, 2, -1, -1, -1, 3 };
 
 		EMIT_NEW_X86_LEA (cfg, ins, array_reg, index2_reg, fast_log2 [size], MONO_STRUCT_OFFSET (MonoArray, vector));
-		ins->klass = m_class_get_element_class (klass);
+		ins->klass = klass;
```
Comment (lambdageek) [inline mono/mini/method-to-ir.c]: How precise does the class have to be in IR?  `klass` here is some array type, right?  So this should be a reference (or maybe ptr) to the array element type - or would that end up generating worse code down the line?

Comment (vargaz) [reply mono/mini/method-to-ir.c]: Its mostly used for verification.

File: mono/mini/arrays.cs [hunk 2]
```diff
@@ -840,8 +840,30 @@ private unsafe static void WritePtr (FooStruct *val, out FooStruct* ptr)
...
+		var f = new JaggedClass ();
+
+		ref int[] r = ref f.a[1];
+
+		return r [0];
+	}
 
+	public static int bar (ref int[] x) {
+		return x [0];
+	}
```
Comment (lewurm) [inline mono/mini/arrays.cs]: `bar ()` is not used

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**